tests/mem_safety/test_stg_st2g_code_generation.wast

```diff
@@ -0,0 +1,294 @@
+  ;; st2g
+  ;; st2g
+  ;; st2g
+  ;; st2g
+  ;; st2g
+  ;;
+  ;; st2g
+  ;; st2g
+  ;; st2g
+  ;; st2g
+  ;; stg
+  ;;
+  (func (export "just_less_than_two_threshholds") (result i32)
+        ;; int x[76]; (1 int is 4 bytes, 4 * 76 bytes = 2*160 - 16 = 304 bytes total)
+        (local $ptr i64) ;; declare variable ptr
+
+        ;; tag memory region
+        (i64.const 0) ;; index (offset/ptr in wasm linear memory)
+        (i64.const 304) ;; size of region we want to tag
+        (segment.stack_new) ;; returns tagged_index on stack
+        (local.tee $ptr) ;; store tagged_index into $ptr
+
+        ;; x[0] = 42;
+        (i32.const 42) ;; push 42 on stack
+        (i32.store) ;; store value 42 into $ptr (ptr[0] = 42)
+
+        ;; push ptr[0] onto stack
+        (local.get $ptr) ;; push $ptr onto stack
+        (i32.load) ;; push back ptr[0] onto stack
+
+        ;; untag memory region
+        (local.get $ptr) ;; push $ptr value to stack
+        (i64.const 0) ;; index
+        (i64.const 304) ;; size
+        (segment.stack_free)
+
+        ;; return x[0];
+        (return)
+  )
+
+
+
+  ;; Test 4:
+  ;;
+  ;; Equivalent C function:
+  ;;
+  ;; int* foo() {
+  ;;   int[124];
+  ;;   x[0] = 42;
+  ;;   return x[0];
+  ;; }
+  ;;
+  ;; What stg/st2g instructions should be generated:
+  ;; for _ in 0..3:
+  ;;   st2g
+  ;;   st2g
+  ;;   st2g
+  ;;   st2g
+  ;;   st2g
+  ;; stg
+  ;;
+  (func (export "multiple_threshhold_loops") (result i32)
+        ;; int x[124]; (1 int is 4 bytes, 4 * 124 bytes = 496 bytes total)
+        (local $ptr i64) ;; declare variable ptr
+
+        ;; tag memory region
+        (i64.const 0) ;; index (offset/ptr in wasm linear memory)
+        (i64.const 496) ;; size of region we want to tag
+        (segment.stack_new) ;; returns tagged_index on stack
+        (local.tee $ptr) ;; store tagged_index into $ptr
+
+        ;; x[0] = 42;
+        (i32.const 42) ;; push 42 on stack
+        (i32.store) ;; store value 42 into $ptr (ptr[0] = 42)
+
+        ;; push ptr[0] onto stack
+        (local.get $ptr) ;; push $ptr onto stack
+        (i32.load) ;; push back ptr[0] onto stack
+
+        ;; untag memory region
+        (local.get $ptr) ;; push $ptr value to stack
+        (i64.const 0) ;; index
+        (i64.const 496) ;; size
+        (segment.stack_free)
+
+        ;; return x[0];
+        (return)
+  )
+
+
+
+  ;; Test 5:
+  ;;
+  ;; Equivalent C function:
+  ;;
+  ;; int* foo(int i) {
+  ;;   int[i];
+  ;;   x[0] = 42;
+  ;;   return x[0];
+  ;; }
+  ;;
+  ;; What stg/st2g instructions should be generated:
+  ;; Completely dynamic, no static optimizations
+  ;;
+  (func (export "non_constant_dynamic_length") (param $i i64) (result i32)
+        ;; int x[i]; (1 int is 4 bytes, 4 * i bytes total)
+        (local $ptr i64) ;; declare variable ptr
+
+        ;; tag memory region
+        (i64.const 0) ;; index (offset/ptr in wasm linear memory)
+        (local.get $i) ;; size of region we want to tag
```

Comment:
I think you need a left shift by 2 here, since `i` declares the number of integers in the array. If you pass in `i = 1`, only one byte is allocated, which is not enough to load an `i32`. Same below when calling `segment.stack_free`.